check_diff/tests/check_diff.rs

```diff
@@ -0,0 +1,80 @@
+use check_diff::{check_diff, compile_rustfmt, search_for_rs_files, CheckDiffError};
+use std::fs::File;
+use tempfile::Builder;
+
+#[test]
+fn search_for_files_correctly_non_nested() -> Result<(), Box<dyn std::error::Error>> {
+    let dir = Builder::new().tempdir_in("").unwrap();
+    let file_path = dir.path().join("test.rs");
+    let _tmp_file = File::create(file_path)?;
+
+    let iter = search_for_rs_files(dir.path());
+
+    let mut count = 0;
+    for _ in iter {
+        count += 1;
+    }
+
+    assert_eq!(count, 1);
+
+    Ok(())
+}
+
+#[test]
+fn search_for_files_correctly_nested() -> Result<(), Box<dyn std::error::Error>> {
+    let dir = Builder::new().tempdir_in("").unwrap();
+    let file_path = dir.path().join("test.rs");
+    let _tmp_file = File::create(file_path)?;
+
+    let nested_dir = Builder::new().tempdir_in(dir.path()).unwrap();
+    let nested_file_path = nested_dir.path().join("nested.rs");
+    let _ = File::create(nested_file_path)?;
+
+    let iter = search_for_rs_files(dir.path());
+
+    let mut count = 0;
+    for _ in iter {
+        count += 1;
+    }
+
+    assert_eq!(count, 2);
+
+    Ok(())
+}
+
+#[test]
+fn check_diff_test() -> Result<(), CheckDiffError> {
+    let tmp_dir = Builder::new().tempdir_in("").unwrap();
+    let runners = compile_rustfmt(
+        tmp_dir.path(),
+        "https://github.com/rust-lang/rustfmt".to_string(),
+        "rustfmt-1.4.32".to_string(),
+        None,
+    )?;
```

Comment:
I think this is a cleaver way to test `check_diff`. But, the downside is that it requires network access to actually test the `check_diff` function. I think it would be great if we didn't need to build either of the rustfmt binaries in order to do so. As I mentioned in an earlier comment we could create a trait named something like `CodeFormatter`, that matches the `format_code` signature we've already defined for `RustfmtRunner`. Then the `CheckDiffRunner` could be made generic over the `CodeFormatter`, which will make testing `check_diff` a lot simpler.

```rust
pub struct CheckDiffRunner<F, S> {
    feature_runner: F,
    src_runner: S,
}
```

```rust
impl<F, S> CheckDiffRunner<F, S> {
    pub fn new(feature_runner: F, src_runner: S) -> Self {
        Self { feature_runner, src_runner }
    }
}

impl<F, S> CheckDiffRunner<F, S>
Where
    F: CodeFormatter,
    S: CodeFormatter,
{
    /// Creates a diff generated by running the source and feature binaries on the same file path
    pub fn create_diff(
        &self,
        path: &Path,
        additional_configs: &Option<Vec<String>>,
    ) -> Result<Diff, CheckDiffError> {
        let code = std::fs::read_to_string(path)?;
        let src_format = self.src_runner.format_code(&code, additional_configs)?;
        let feature_format = self.feature_runner.format_code(&code, additional_configs)?;
        Ok(Diff {
            src_format,
            feature_format,
        })
    }
}
```

Then the test(s) could be something like this:

```rust
#[test]
fn check_diff_test_no_formatting_differences() {
    let runner = CheckDiffRunner::new(
        /* `CodeFormatter` implementations that agree on the formatting */
    );
    let errors = check_diff(None, runner, dir.path());
    assert!(errors == 0);
}

#[test]
fn check_diff_test_with_formatting_differences() {
    let runner = CheckDiffRunner::new(
        /* `CodeFormatter` implementations that don't agree on the formatting */
    );
    let errors = check_diff(None, runner, dir.path());
    assert!(errors >= 1);
}
```